From 103b02bd32ed65174b16af02215807d8d0e133a6 Mon Sep 17 00:00:00 2001
From: Andy Hu <andy.hu@starfivetech.com>
Date: Tue, 16 Aug 2022 21:58:00 +0800
Subject: [PATCH] codaj12: fix building issue

---
 codaj12/codaj12_buildroot.mak | 31 ++++++++++++++++++++++++++++++-
 1 file changed, 30 insertions(+), 1 deletion(-)

diff --git a/codaj12/codaj12_buildroot.mak b/codaj12/codaj12_buildroot.mak
index bffe02d..6833f09 100644
--- a/codaj12/codaj12_buildroot.mak
+++ b/codaj12/codaj12_buildroot.mak
@@ -53,7 +53,8 @@ ifeq ("$(BUILD_CONFIGURATION)", "EmbeddedLinux")
 endif
 
 ifeq ("$(BUILD_CONFIGURATION)", "EmbeddedRiscvLinux")
-    CROSS_CC_PREFIX = riscv64-buildroot-linux-gnu-
+#    CROSS_CC_PREFIX = riscv64-buildroot-linux-gnu-
+    CROSS_CC_PREFIX =
     PLATFORM        = riscvlinux
 endif
 CC  = $(CROSS_CC_PREFIX)gcc
@@ -146,3 +147,31 @@ obj/%.o: %.c $(MAKEFILE)
 lint:
 	"$(LINT_HOME)/flint" -i"$(LINT_HOME)" $(DEFINES) $(INCLUDES) $(VPATH2) linux_std.lnt $(HAPS_RULE) $(NONOS_RULE)  $(SOURCES_COMMON)
 
+TARGET_DIR=../target
+STAGING_DIR=$(TARGET_DIR)
+INSTALL=install
+
+install:
+	@echo install ...
+	$(INSTALL) -D -m 0777 jdi/linux/driver/load.sh $(TARGET_DIR)/root/codaj12/load.sh
+	$(INSTALL) -D -m 0777 jdi/linux/driver/unload.sh $(TARGET_DIR)/root/codaj12/unload.sh
+	$(INSTALL) -D -m 0644 libcodadec.so $(TARGET_DIR)/usr/lib/libcodadec.so
+	@echo "install STAGING_DIR ..."
+	$(INSTALL) -D -m 0644 jpuapi/jpuapi.h                  $(STAGING_DIR)/usr/include/codaj12/jpuapi/jpuapi.h
+	$(INSTALL) -D -m 0644 jpuapi/jpuapifunc.h              $(STAGING_DIR)/usr/include/codaj12/jpuapi/jpuapifunc.h
+	$(INSTALL) -D -m 0644 jpuapi/regdefine.h               $(STAGING_DIR)/usr/include/codaj12/jpuapi/regdefine.h
+	$(INSTALL) -D -m 0644 jpuapi/jpuconfig.h               $(STAGING_DIR)/usr/include/codaj12/jpuapi/jpuconfig.h
+	$(INSTALL) -D -m 0644 jpuapi/jputypes.h                $(STAGING_DIR)/usr/include/codaj12/jpuapi/jputypes.h
+	$(INSTALL) -D -m 0644 jpuapi/jputable.h                $(STAGING_DIR)/usr/include/codaj12/jpuapi/jputable.h
+	$(INSTALL) -D -m 0644 sample/helper/cnm_fpga.h         $(STAGING_DIR)/usr/include/codaj12/sample/helper/cnm_fpga.h
+	$(INSTALL) -D -m 0644 sample/helper/platform.h         $(STAGING_DIR)/usr/include/codaj12/sample/helper/platform.h
+	$(INSTALL) -D -m 0644 sample/helper/yuv_feeder.h       $(STAGING_DIR)/usr/include/codaj12/sample/helper/yuv_feeder.h
+	$(INSTALL) -D -m 0644 sample/helper/datastructure.h    $(STAGING_DIR)/usr/include/codaj12/sample/helper/datastructure.h
+	$(INSTALL) -D -m 0644 sample/helper/jpulog.h           $(STAGING_DIR)/usr/include/codaj12/sample/helper/jpulog.h
+	$(INSTALL) -D -m 0644 sample/main_helper.h             $(STAGING_DIR)/usr/include/codaj12/sample/main_helper.h
+	$(INSTALL) -D -m 0644 jdi/linux/driver/jpu.h           $(STAGING_DIR)/usr/include/codaj12/jdi/linux/driver/jpu.h
+	$(INSTALL) -D -m 0644 jdi/linux/driver/jmm.h           $(STAGING_DIR)/usr/include/codaj12/jdi/linux/driver/jmm.h
+	$(INSTALL) -D -m 0644 jdi/jdi.h                        $(STAGING_DIR)/usr/include/codaj12/jdi/jdi.h
+	$(INSTALL) -D -m 0644 jdi/mm.h                         $(STAGING_DIR)/usr/include/codaj12/jdi/mm.h
+	$(INSTALL) -D -m 0644 config.h                         $(STAGING_DIR)/usr/include/codaj12/config.h
+
-- 
2.36.0

