From 14143fcf2c4561905c11cae22314bb5dfdbe171c Mon Sep 17 00:00:00 2001
From: Andy Hu <andy.hu@starfivetech.com>
Date: Wed, 17 Aug 2022 11:14:38 +0800
Subject: [PATCH] omx-il: fix building issue

---
 omx-il/Makefile | 29 ++++++++++++++++++++++-------
 1 file changed, 22 insertions(+), 7 deletions(-)

diff --git a/omx-il/Makefile b/omx-il/Makefile
index 150c4d5..a51de48 100644
--- a/omx-il/Makefile
+++ b/omx-il/Makefile
@@ -10,7 +10,8 @@ BUILD_CONFIGURATION := EmbeddedRiscvLinux
 PRODUCT := OMX_IL
 
 ifeq ("$(BUILD_CONFIGURATION)", "EmbeddedRiscvLinux")
-    CROSS_CC_PREFIX = riscv64-buildroot-linux-gnu-
+#    CROSS_CC_PREFIX = riscv64-buildroot-linux-gnu-
+    CROSS_CC_PREFIX =
     PLATFORM        = riscvlinux
 endif
 CC  = $(CROSS_CC_PREFIX)gcc
@@ -18,9 +19,13 @@ CXX = $(CROSS_CC_PREFIX)g++
 LINKER=$(CC)
 AR  = $(CROSS_CC_PREFIX)ar
 
+TARGET_DIR=../target
+STAGING_DIR=$(TARGET_DIR)
+INSTALL=install
+
 DEFINES = -DUSE_FEEDING_METHOD_BUFFER
 INCLUDES = -I./include/khronos -I./core -I./component/video/wave5/common -I./component/video/wave4 -I./component/helper
-$(warning "the value of LOCAL_PATH is$(STAGING_DIR)")
+$(warning "the value of LOCAL_PATH is $(STAGING_DIR)")
 INCLUDES += -I$(STAGING_DIR)/usr/include/
 INCLUDES += -I$(STAGING_DIR)/usr/include/wave420l/
 INCLUDES += -I$(STAGING_DIR)/usr/include/wave420l/vpuapi/
@@ -29,12 +34,13 @@ INCLUDES += -I$(STAGING_DIR)/usr/include/wave420l/sample/helper/
 
 #mjpeg
 INCLUDES += -I./component/image/common
+INCLUDES += -I../out/include
 
 CFLAGS  += -g -I. $(INCLUDES) $(DEFINES) $(PLATFORM_FLAGS)
 CFLAGS  += -fpic
 ARFLAGS += cru
-LDFLAGS = -ldl -lpthread -Wl,--fatal-warning
-LDFLAGS_FFMPEG = -lavformat -lavcodec -lavutil -lswresample
+LDFLAGS = -L./  -ldl -lpthread -Wl,--fatal-warning
+LDFLAGS_FFMPEG = -L../out/lib -lavformat -lavcodec -lavutil -lswresample 
 
 OBJDIR=obj
 ALLOBJS=*.o
@@ -86,18 +92,27 @@ omx-il: CREATE_DIR $(OBJECTPATHS_COMMON)
 test: video-dec-test video-enc-test mjpeg-dec-test
 
 video-dec-test: CREATE_DIR $(OBJECTPATHS_DEC_TEST)
-	$(CC) -o video_dec_test $(LDFLAGS) $(LDFLAGS_FFMPEG) -lsf-omx-il $(OBJECTPATHS_DEC_TEST) -L./
+	$(CC) -o video_dec_test $(OBJECTPATHS_DEC_TEST) $(LDFLAGS) $(LDFLAGS_FFMPEG) -lsf-omx-il 
 
 video-enc-test: CREATE_DIR $(OBJECTPATHS_ENC_TEST)
-	$(CC) -o video_enc_test $(LDFLAGS) $(LDFLAGS_FFMPEG) -lsf-omx-il $(OBJECTPATHS_ENC_TEST) -L./
+	$(CC) -o video_enc_test $(OBJECTPATHS_ENC_TEST) -L../out/lib -L./ -lsf-omx-il $(LDFLAGS_FFMPEG) $(LDFLAGS)
 
 mjpeg-dec-test: CREATE_DIR $(OBJECTPATHS_MJPEG_DEC_TEST)
-	$(CC) -g -o mjpeg_dec_test $(LDFLAGS) $(LDFLAGS_FFMPEG) -lsf-omx-il $(OBJECTPATHS_MJPEG_DEC_TEST) -L./
+	$(CC) -g -o mjpeg_dec_test $(OBJECTPATHS_MJPEG_DEC_TEST) $(LDFLAGS) $(LDFLAGS_FFMPEG) -lsf-omx-il -L./
 
 clean:
+	rm -rf obj/
+	rm -rf libsf-omx-il.so video-dec-test video-enc-test mjpeg-dec-test
 
 CREATE_DIR:
 	-mkdir -p $(OBJDIR)
 
 obj/%.o: %.c $(MAKEFILE)
 	$(CC) $(CFLAGS) -Wall -Werror -c $< -o $@ -MD -MF $(@:.o=.dep)
+
+install:
+	@echo "install omx-il to STAGING_DIR ..."
+	$(INSTALL) -m 0644 libsf-omx-il.so $(STAGING_DIR)/usr/lib/libsf-omx-il.so
+	ln -sf libsf-omx-il.so $(STAGING_DIR)/usr/lib/libOMX_Core.so
+	$(INSTALL) -d include/khronos $(STAGING_DIR)/usr/include/omx-il
+	$(INSTALL) -m 0644 include/khronos/* $(STAGING_DIR)/usr/include/omx-il
-- 
2.36.0

